content/blog/2019-10-08-remote-read-meets-streaming.md

```diff
@@ -0,0 +1,313 @@
+
+The remote read protocol was extended in a backward and forward compatible way:
+
+* Prometheus before v2.13.0 will safely ignore the `accepted_response_types` field provided by newer clients and assume `SAMPLED` mode.
+* Prometheus after v2.13.0 will default to the `SAMPLED` mode for older clients that don't provide the `accepted_response_types` parameter.
+
+## Usage
+
+To use the new, streamed remote read in Prometheus v2.13.0, a 3rd party system has to add  `accepted_response_types = [STREAMED_XOR_CHUNKS]` to the request.
+
+Then Prometheus will stream `ChunkedReadResponse` instead of old message. Each `ChunkedReadResponse` message is 
+following varint size and fixed size bigendian uint32 for CRC32 Castagnoli checksum.
+
+For Go it is recommended to use our [ChunkedReader](https://github.com/prometheus/prometheus/blob/48b2c9c8eae2d4a286d8e9384c2918aefd41d8de/storage/remote/chunked.go#L103)
+ to read directly from the stream.   
+
+Note that `storage.remote.read-sample-limit` flag is no longer working for `STREAMED_XOR_CHUNKS`.
+`storage.remote.read-concurrent-limit` works as previously.
+
+There also new option `storage.remote.read-max-bytes-in-frame` which controls the maximum size of each message. It is advised
+to keep it 1MB as the default as it is recommended by Google to keep protobuf message [not larger than 1MB](https://developers.google.com/protocol-buffers/docs/techniques#large-data).
+
+As mentioned before, [Thanos](https://thanos.io) gains a lot with this improvement. We added support for the new remote read  in `v0.7.0`, so this version or any newer, 
+will use streamed remote read automatically whenever Prometheus 2.13 or newer is used next to the Thanos sidecar.
+
+## Next Steps
+
+Release 2.13.0 introduces extended remote read and Prometheus server side implementation, However at the moment of writing
+there are still few items to do in order to fully get advantage from the extended remote read protocol:
+
+* Support for client side of Prometheus remote read: [In progress](https://github.com/prometheus/prometheus/issues/5926) 
+* Avoid reencoding of chunks for blocks during remote read: [In progress](https://github.com/prometheus/prometheus/pull/5882)
+
+## Summary
+
+To sum up, the main benefits of chunked, streaming of remote read are:
+
+* Both client and server are capable to use **constant memory size and per each request**. This is because Prometheus now
+at maximum buffers just a single small frame instead of the whole response during remote read. This massively helps with
+capacity planning, especially for non-compressible resource like memory.
+* Prometheus server does not need to decode chunks to raw samples anymore during remote read. The same for client side for
+encoding, **if** the system is reusing native TSDB XOR compression (like Thanos does). 
+
+As always, if we you have any issues or feedback, feel free to submit a ticket on GitHub or reach us on the mailing list.
```

Comment:
```suggestion
As always, if you have any issues or feedback, feel free to submit a ticket on GitHub or ask questions on the mailing list.
```
feels with less responsible to provide support

Reply:
Sorry what do you mean?